File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/relicRecovery2017/AutonomousRoutesOp.java
package org.firstinspires.ftc.teamcode.relicRecovery2017;

import com.qualcomm.robotcore.eventloop.opmode.Autonomous;
import com.qualcomm.robotcore.eventloop.opmode.Disabled;
import com.qualcomm.robotcore.eventloop.opmode.LinearOpMode;

import org.firstinspires.ftc.robotcore.external.Telemetry;
@Disabled
@Autonomous(name="autonomous routes")
public class AutonomousRoutesOp extends LinearOpMode {

    AutonomousHardware robot = new AutonomousHardware();
    Telemetry.Item pathTelemetry = null;
    Telemetry.Item targetTelemetry = null;

    @Override
    public void runOpMode() throws InterruptedException {

        robot.init(hardwareMap, telemetry);
        robot.setOpMode(this);

        int selectedOption = 1;
        boolean pressed = false;

        while (!gamepad1.a && !isStopRequested()) {
            if (!pressed && gamepad1.dpad_up && selectedOption < 3) {
                selectedOption++;
            } else if (!pressed && gamepad1.dpad_down && selectedOption > 1) {
                selectedOption--;
            }
            pressed = gamepad1.dpad_up || gamepad1.dpad_down;

            telemetry.addData("Route", selectedOption);
            telemetry.update();
        }

        telemetry.addData("status" , "ready to start");
        telemetry.update();

        waitForStart();

        telemetry.setAutoClear(false);

        telemetry.addData("status", "running");

        pathTelemetry = telemetry.addData("path", "init");
        targetTelemetry = telemetry.addData("target", "init");

        try {
            switch (selectedOption) {
                case 1: route1(); break;
                case 2: route2(); break;
                case 3: route3();break;
            }
        } catch (Exception exception) {}
    }

    void route1() throws Exception {
        for (int i=1; i <= 4; i++) {
            double xAxis = 0.5*(((i%2) ^ 1) * (-(i/4) | 1));
            double yAxis = 0.5*((i%2) * (-(i/3) | 1));
            pathTelemetry.setValue("drive");
            targetTelemetry.setValue("("+xAxis+", "+yAxis+")");
            robot.driveForSec(xAxis, yAxis, 1);
        }
        robot.moveRobot();
    }

    void route2() throws Exception {
        for (int i=1; i <= 2; i++) {
            pathTelemetry.setValue("drive forward" + i);
            targetTelemetry.setValue("(0, 0.5)");
            robot.driveForSec(0, 0.1, 1);

            robot.moveRobot();

            pathTelemetry.setValue("turn " + i);
            targetTelemetry.setValue(90*i);
            robot.turnToHeading(90.0*i, 0.2, 0, 0.1);
        }

        for (int i=3; i <= 4; i++) {
            pathTelemetry.setValue("drive forward " + i);
            targetTelemetry.setValue("(0, 0.5)");
            robot.driveForSec(0, 0.1, 1);

            robot.moveRobot();

            pathTelemetry.setValue("turn " + i);
            targetTelemetry.setValue(-180 + 90*i);
            robot.turnToHeading(-180.0 + 90*i, 0.2, 0, 0.1);
        }
        robot.moveRobot();
    }

    void route3() throws Exception {
        pathTelemetry.setValue("turn");
        targetTelemetry.setValue("0 (0.0, 1) 0.9");
        robot.moveRobot(0, 1, 0.9);
        robot.turnToHeading(0, 0.9, 0, 1, 1);

        robot.moveRobot();
    }
}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/kotlinTestCode/AutonomousRoutesOp.kt
package org.firstinspires.ftc.teamcode.kotlinTestCode

import com.cout970.vector.impl.Vector2d
import com.qualcomm.robotcore.eventloop.opmode.Autonomous
import com.qualcomm.robotcore.eventloop.opmode.LinearOpMode
import org.firstinspires.ftc.robotcore.external.Telemetry

@Autonomous(name = "Auto Routes", group = "auto tests")
class AutonomousRoutesOp:  LinearOpMode(){

    var robot: AutonomousHardware = AutonomousHardware()
    var pathTelementry: Telemetry.Item? = null
    var targetTelementry: Telemetry.Item? = null

    override fun runOpMode() {

        robot.init(hardwareMap, telemetry)

        var selectedOption = 1
        var pressed = false

        while (!gamepad1.a && !isStopRequested) {
            if (!pressed && gamepad1.dpad_up && selectedOption < 3) {
                selectedOption++
            } else if (!pressed && gamepad1.dpad_down && selectedOption > 1) {
                selectedOption--
            }
            pressed = gamepad1.dpad_up || gamepad1.dpad_down

            telemetry.addData("Route", selectedOption)
            telemetry.update()
        }

        telemetry.addData("status", "ready to start")
        telemetry.update()

        waitForStart()

        telemetry.isAutoClear = false

        telemetry.addData("status", "running");

        pathTelementry = telemetry.addData("path","init")
        targetTelementry = telemetry.addData("target", "init")

        telemetry.addData("heading", {robot.currentHeading})

        robot.loopFunction = {autoLoop()}

        try {
            when (selectedOption) {
                1 -> route1()
                2 -> route2()
                3 -> route3()
            }
        } catch (exception: Exception) {}
    }

    fun autoLoop() {
        telemetry.update()
        if (!opModeIsActive()) {
            throw Exception();
        }
    }

    fun route1() {
        for (i in 1..4) {
            val target = Vector2d(0.5*(((i%2) xor 1 ) * (-(i/4) or 1)), 0.5*((i%2) * (-(i/3) or 1)))
            pathTelementry?.setValue("drive")
            targetTelementry?.setValue("("+target.x+", "+target.y+")")
            robot.driveForSecs(target, 1.0)
        }
        robot.moveRobot()
    }

    fun route2() {
        for (i in 1..2) {
            pathTelementry?.setValue("drive forward $i")
            targetTelementry?.setValue("(0, 0.5)")
            robot.driveForSecs(Vector2d(0.0, 0.5), 1.0)

            robot.moveRobot()

            pathTelementry?.setValue("turn $i")
            targetTelementry?.setValue(90*i)
            robot.turnToHeading(90.0*i, 0.2, Vector2d(0.0, 0.1))
        }

        for (i in 1..2) {
            pathTelementry?.setValue("drive forward $i")
            targetTelementry?.setValue("(0, 0.5)")
            robot.driveForSecs(Vector2d(0.0, 0.5), 1.0)

            robot.moveRobot()

            pathTelementry?.setValue("turn $i")
            targetTelementry?.setValue(-180 + 90*i)
            robot.turnToHeading(-180.0 + 90*i, 0.2, Vector2d(0.0, 0.1))
        }
        robot.moveRobot()
    }

    fun route3() {
        pathTelementry?.setValue("turn")
        targetTelementry?.setValue("0 (0.0, 1.0) 0.9")
        robot.moveRobot(0.0, 1.0, 0.9)
        robot.turnToHeading(0.0, 0.9, Vector2d(0.0, 1.0), 1)

        robot.moveRobot()
    }
}